spring-kafka/src/main/java/org/springframework/kafka/annotation/EnableRetryTopic.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.kafka.config.KafkaBackOffManagerConfigurationSupport;
+import org.springframework.kafka.config.RetryTopicConfigurationSupport;
+
+/**
+ * Enables the non-blocking topic-based delayed retries feature. To be used in
+ * {@link Configuration Configuration} classes as follows:
+ * <pre class="code">
+ *
+ * &#064;EnableRetryTopic
+ * &#064;Configuration
+ * public class AppConfig {
+ *
+ * 	&#064;Bean
+ * 	public RetryTopicConfiguration myRetryTopicConfiguration(KafkaTemplate kafkaTemplate) {
```

Comment:
I this optional?
I thought something like `@EnableRetryTopic` should be fair enough for some default values...

Reply:
Haven't thought of that, sounds interesting. I can think this through, but at first it seems to me that it might not be so simple - currently if we have a greedy configuration that accepts all topics, it might impede users from having specific configurations for specific topics.

We could either make a greedy, general configuration possible (that's something I'd been thinking of anyway) with users only providing the specific ones, or maybe there's a way we can add a bean conditionally on not having a user provided one? I see all `@ConditionalOn` goodies are part of Boot and we don't have it here.

WDYT?

Reply:
The `@EnableRetryTopic` may come with attributes to configure something global or for default values...

Reply:
Sure, that sounds interesting too. But I wonder, Andy Wilkinson from the Boot team pinged me earlier today to look into having auto configuration for the retryable topics in RC1 next week, not sure you saw that. Wouldn't those two approaches maybe conflict a bit?

Users shouldn't need `@EnableRetryTopic` to make the feature work with auto configuration, since boot won't have access to it until 3.0, but I think they still should be able to use it if, for example, they want to inject a feature bean such as the `DestinationTopicResolver`, as we've discussed.

WDYT?

Reply:
Correct. Independently of the solution here, the auto-configuration should give end-user whatever is needed to be done by default or from the configuration properties.

I think I see your pain and perhaps we need just to deffer this new vision until `3.0` and let you concentrate on the auto-configuration for upcoming Spring Boot `2.7`.
Note: Gary doesn't have intention to release `spring-kafka-2.9.0` for that Spring Boot. 
So, whatever we try to achieve here cannot benefit from the auto-configuration for a while.
I mean I still doubt this can make it smoothly to `2.8.x`...

Reply:
I'm ok with working with either 2.x or 3.0, or both. What I don't think I'm really comfortable with is adding features to 2.x with the current configuration system in place, since it'll only bring confusion and pain for our users. Does that make sense, or maybe I am missing something?

So, the way I see it, we can insist on making this backwards compatible with 2.x, or maybe come up with simple, less-than-perfect patches to keep the 2.x ball rolling until it's a bit closer to the 3.0 release. Or we can just leave everything new to 3.0 and be done with 2.x as far as retryable topics go, except of course for bug fixes. I'm ok with both approaches.

WDYT? If you have any other ideas I'm all ears too.

Thanks!

Reply:
As a note, for now all auto-configuration will do is create a global `RetryTopicConfiguration` bean with a few parameters - no infrastructure involved, and I guess not too much work to be done there too. While it might help getting the feature in the hands of more users, it shouldn't really help with any of this infrastructure problem. Of course, if you have any different ideas please let me know.

Reply:
So, maybe what we might be able to do is having `@EnableRetryTopic` provide a base configuration if it doesn't find any other `RetryTopicConfiguration` beans - this way if the user has provided any other configuration, be it through auto-configuration or via a `@Configuration` class, it will back off, otherwise, this default configuration kicks in, and we could add a couple configuration properties there just as we do in auto configuration.

How does that sound? I like the idea since it allows users to have a declarative way of using the feature without needing to mess around with yml files.

Reply:
I think we may have had some kind of misunderstanding here 😄

> Wouldn't those two approaches maybe conflict a bit?

What I meant is that auto configuration might conflict with the idea of `@EnableRetryTopic` providing a default configuration - not with this `Support` approach as a whole. Is that what you understood? If so, sorry for not being clearer.

> I think I see your pain and perhaps we need just to deffer this new vision until 3.0 and let you concentrate on the auto-configuration for upcoming Spring Boot 2.7.

So I didn't really understand why we might defer this to 3.0. If that's only because of focusing on auto-configuration, I don't think there's any conflict between them and I can do both.

Please let me know if I understood any of this incorrectly. Either way, I should submit the changes we've discussed sometime today, and I'll also begin looking at auto-configuration.

Thanks again.